File: modules/core/event.go
package core

import (
	"context"
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// AgentEvent interface
// ---------------------------------------------------------------------------

// AgentEvent is the universal interface for all events emitted by an agent.
type AgentEvent interface {
	EventType() string
	AgentID() AgentId
	SessionID() SessionId
	Timestamp() time.Time
}

// ---------------------------------------------------------------------------
// BaseEvent
// ---------------------------------------------------------------------------

// BaseEvent is embedded in every concrete event type and provides the four
// fields required by the AgentEvent interface.
type BaseEvent struct {
	Type    string    `json:"type"`
	Agent   AgentId   `json:"agent_id"`
	Session SessionId `json:"session_id"`
	Time    time.Time `json:"timestamp"`
}

// EventType returns the event type string.
func (b BaseEvent) EventType() string { return b.Type }

// AgentID returns the agent that produced this event.
func (b BaseEvent) AgentID() AgentId { return b.Agent }

// SessionID returns the session this event belongs to.
func (b BaseEvent) SessionID() SessionId { return b.Session }

// Timestamp returns the wall-clock time when the event was created.
func (b BaseEvent) Timestamp() time.Time { return b.Time }

// ---------------------------------------------------------------------------
// Concrete event types (11)
// ---------------------------------------------------------------------------

// 1. StartedEvent — agent has started.
type StartedEvent struct{ BaseEvent }

// NewStartedEvent creates a StartedEvent with the current timestamp.
func NewStartedEvent(agentID AgentId, sessionID SessionId) StartedEvent {
	return StartedEvent{
		BaseEvent: BaseEvent{
			Type:    "started",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
	}
}

// 2. CompletedEvent — agent has finished.
type CompletedEvent struct {
	BaseEvent
	Summary string `json:"summary"`
}

// NewCompletedEvent creates a CompletedEvent with a summary.
func NewCompletedEvent(agentID AgentId, sessionID SessionId, summary string) CompletedEvent {
	return CompletedEvent{
		BaseEvent: BaseEvent{
			Type:    "completed",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		Summary: summary,
	}
}

// 3. MessageReceivedEvent — agent received a message.
type MessageReceivedEvent struct {
	BaseEvent
	ContentPreview string `json:"content_preview"`
}

// NewMessageReceivedEvent creates a MessageReceivedEvent with a content preview.
func NewMessageReceivedEvent(agentID AgentId, sessionID SessionId, preview string) MessageReceivedEvent {
	return MessageReceivedEvent{
		BaseEvent: BaseEvent{
			Type:    "message_received",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		ContentPreview: preview,
	}
}

// 4. ToolCallRequestedEvent — agent is requesting a tool call.
type ToolCallRequestedEvent struct {
	BaseEvent
	ToolCall ToolCall `json:"tool_call"`
}

// NewToolCallRequestedEvent creates a ToolCallRequestedEvent.
func NewToolCallRequestedEvent(agentID AgentId, sessionID SessionId, tc ToolCall) ToolCallRequestedEvent {
	return ToolCallRequestedEvent{
		BaseEvent: BaseEvent{
			Type:    "tool_call_requested",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		ToolCall: tc,
	}
}

// 5. ToolCallCompletedEvent — a tool call finished.
type ToolCallCompletedEvent struct {
	BaseEvent
	ToolName   string `json:"tool_name"`
	Success    bool   `json:"success"`
	DurationMs uint64 `json:"duration_ms"`
}

// NewToolCallCompletedEvent creates a ToolCallCompletedEvent.
func NewToolCallCompletedEvent(agentID AgentId, sessionID SessionId, name string, success bool, durMs uint64) ToolCallCompletedEvent {
	return ToolCallCompletedEvent{
		BaseEvent: BaseEvent{
			Type:    "tool_call_completed",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		ToolName:   name,
		Success:    success,
		DurationMs: durMs,
	}
}

// 6. TokenUsageUpdatedEvent — token usage has been updated.
type TokenUsageUpdatedEvent struct {
	BaseEvent
	Usage TokenUsage `json:"usage"`
}

// NewTokenUsageUpdatedEvent creates a TokenUsageUpdatedEvent.
func NewTokenUsageUpdatedEvent(agentID AgentId, sessionID SessionId, usage TokenUsage) TokenUsageUpdatedEvent {
	return TokenUsageUpdatedEvent{
		BaseEvent: BaseEvent{
			Type:    "token_usage_updated",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		Usage: usage,
	}
}

// 7. StreamChunkEvent — a chunk of streaming content.
type StreamChunkEvent struct {
	BaseEvent
	Content string `json:"content"`
}

// NewStreamChunkEvent creates a StreamChunkEvent.
func NewStreamChunkEvent(agentID AgentId, sessionID SessionId, content string) StreamChunkEvent {
	return StreamChunkEvent{
		BaseEvent: BaseEvent{
			Type:    "stream_chunk",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		Content: content,
	}
}

// 8. ErrorEvent — an error occurred.
type ErrorEvent struct {
	BaseEvent
	ErrorMessage string `json:"error_message"`
}

// NewErrorEvent creates an ErrorEvent.
func NewErrorEvent(agentID AgentId, sessionID SessionId, errMsg string) ErrorEvent {
	return ErrorEvent{
		BaseEvent: BaseEvent{
			Type:    "error",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		ErrorMessage: errMsg,
	}
}

// 9. GoalPhaseChangedEvent — goal phase has changed.
type GoalPhaseChangedEvent struct {
	BaseEvent
	Phase string `json:"phase"`
	Cycle uint32 `json:"cycle"`
}

// NewGoalPhaseChangedEvent creates a GoalPhaseChangedEvent.
func NewGoalPhaseChangedEvent(agentID AgentId, sessionID SessionId, phase string, cycle uint32) GoalPhaseChangedEvent {
	return GoalPhaseChangedEvent{
		BaseEvent: BaseEvent{
			Type:    "goal_phase_changed",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		Phase: phase,
		Cycle: cycle,
	}
}

// 10. GoalTaskCompletedEvent — a task within a goal has completed.
type GoalTaskCompletedEvent struct {
	BaseEvent
	TaskID          string `json:"task_id"`
	TaskDescription string `json:"task_description"`
	Success         bool   `json:"success"`
}

// NewGoalTaskCompletedEvent creates a GoalTaskCompletedEvent.
func NewGoalTaskCompletedEvent(agentID AgentId, sessionID SessionId, taskID, desc string, success bool) GoalTaskCompletedEvent {
	return GoalTaskCompletedEvent{
		BaseEvent: BaseEvent{
			Type:    "goal_task_completed",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		TaskID:          taskID,
		TaskDescription: desc,
		Success:         success,
	}
}

// 11. GoalCycleCompleteEvent — a goal cycle has completed.
type GoalCycleCompleteEvent struct {
	BaseEvent
	Cycle              uint32 `json:"cycle"`
	TasksCompleted     uint32 `json:"tasks_completed"`
	TasksFailed        uint32 `json:"tasks_failed"`
	VerificationPassed bool   `json:"verification_passed"`
}

// NewGoalCycleCompleteEvent creates a GoalCycleCompleteEvent.
func NewGoalCycleCompleteEvent(agentID AgentId, sessionID SessionId, cycle, completed, failed uint32, verified bool) GoalCycleCompleteEvent {
	return GoalCycleCompleteEvent{
		BaseEvent: BaseEvent{
			Type:    "goal_cycle_complete",
			Agent:   agentID,
			Session: sessionID,
			Time:    time.Now().UTC(),
		},
		Cycle:              cycle,
		TasksCompleted:     completed,
		TasksFailed:        failed,
		VerificationPassed: verified,
	}
}

// ---------------------------------------------------------------------------
// EventHandler interface
// ---------------------------------------------------------------------------

// EventHandler processes events. Implementations must be safe for concurrent use
// when registered with an EventBus.
type EventHandler interface {
	Handle(ev AgentEvent) error
	Name() string
}

// ---------------------------------------------------------------------------
// EventHandlerFunc adapter
// ---------------------------------------------------------------------------

// EventHandlerFunc wraps a function to satisfy the EventHandler interface.
type EventHandlerFunc struct {
	name string
	fn   func(AgentEvent) error
}

// NewEventHandlerFunc creates an EventHandlerFunc with the given name and function.
func NewEventHandlerFunc(name string, fn func(AgentEvent) error) EventHandlerFunc {
	return EventHandlerFunc{name: name, fn: fn}
}

// Handle invokes the wrapped function.
func (h EventHandlerFunc) Handle(ev AgentEvent) error { return h.fn(ev) }

// Name returns the handler name.
func (h EventHandlerFunc) Name() string { return h.name }

// ---------------------------------------------------------------------------
// EventBus
// ---------------------------------------------------------------------------

// EventBus dispatches events to all registered handlers concurrently.
type EventBus struct {
	mu        sync.RWMutex
	subs      map[string]EventHandler
	buffer    int
}

// NewEventBus creates a new EventBus. The buffer parameter controls the channel
// buffer size used internally for each handler during publish.
func NewEventBus(buffer int) *EventBus {
	return &EventBus{
		subs:   make(map[string]EventHandler),
		buffer: buffer,
	}
}

// Subscribe registers a handler. The handler's Name() is used as its unique ID.
// If a handler with the same name already exists it is replaced.
func (b *EventBus) Subscribe(handler EventHandler) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.subs[handler.Name()] = handler
}

// Unsubscribe removes the handler identified by the given ID.
// If no handler with that ID exists, Unsubscribe is a no-op.
func (b *EventBus) Unsubscribe(id string) {
	b.mu.Lock()
	defer b.mu.Unlock()
	delete(b.subs, id)
}

// Publish dispatches ev to every registered handler via a goroutine.
// If ctx is canceled before dispatch begins, no handlers are invoked.
func (b *EventBus) Publish(ctx context.Context, ev AgentEvent) {
	// Check context before acquiring the lock.
	if ctx.Err() != nil {
		return
	}

	b.mu.RLock()
	handlers := make([]EventHandler, 0, len(b.subs))
	for _, h := range b.subs {
		handlers = append(handlers, h)
	}
	b.mu.RUnlock()

	for _, h := range handlers {
		go func(handler EventHandler) {
			_ = handler.Handle(ev)
		}(h)
	}
}

File: packages/core/src/event.ts
/**
 * @module event
 *
 * Agent event types for observability and streaming.
 *
 * Events are emitted during agent execution to provide
 * real-time visibility into the agent's progress:
 * - ModelRequest/ModelResponse: AI API calls
 * - ToolCall/ToolResult: tool invocations
 * - Thinking: reasoning steps
 * - Error: failures and warnings
 * - Status: lifecycle state changes
 */
import type { AgentId, SessionId, TokenUsage } from "./types.js";
import type { ToolCall } from "./message.js";

// ---------------------------------------------------------------------------
// AgentEvent interface
// ---------------------------------------------------------------------------

/**
 * Structural shape that all concrete event types share. Consumers of the
 * EventBus treat events polymorphically through this interface; type
 * discrimination happens via `eventType`. Keeping it minimal avoids forcing
 * every subscriber to import the concrete classes.
 */

export interface AgentEvent {
  eventType: string;
  agentId: AgentId;
  sessionId: SessionId;
  timestamp: Date;
}

// ---------------------------------------------------------------------------
// BaseEvent
// ---------------------------------------------------------------------------

/**
 * Shared base for all 12 concrete event types. Carries the routing triple
 * (type, agent, session) plus a timestamp, and knows how to serialize
 * itself to the snake_case JSON shape used for persistence and transport.
 * Concrete subclasses override {@link toJSON} when they carry extra fields.
 */

export interface BaseEventJSON {
  type: string;
  agent_id: string;
  session_id: string;
  timestamp: string;
}

export class BaseEvent {
  constructor(
    public readonly type: string,
    public readonly agent: AgentId,
    public readonly session: SessionId,
    public readonly time: Date = new Date(),
  ) {}

  get eventType(): string { return this.type; }
  get agentId(): AgentId { return this.agent; }
  get sessionId(): SessionId { return this.session; }
  get timestamp(): Date { return this.time; }

  toJSON(): BaseEventJSON {
    return {
      type: this.type,
      agent_id: this.agent.toJSON(),
      session_id: this.session.toJSON(),
      timestamp: this.time.toISOString(),
    };
  }
}

// ---------------------------------------------------------------------------
// Concrete event types (11)
// ---------------------------------------------------------------------------
// Each event below models a single stage in the agent lifecycle. They are
// intentionally tiny immutable value objects exposing exactly the fields the
// UI / metrics / persistence layers need. Subscribers discriminate by the
// `eventType` string ("started", "tool_call_requested", ...) returned by the
// getter on BaseEvent.

// 1. StartedEvent
export class StartedEvent extends BaseEvent {
  constructor(agentId: AgentId, sessionId: SessionId) {
    super("started", agentId, sessionId);
  }
}

// 2. CompletedEvent
export class CompletedEvent extends BaseEvent {
  constructor(agentId: AgentId, sessionId: SessionId, public readonly summary: string) {
    super("completed", agentId, sessionId);
  }
}

// 3. MessageReceivedEvent
export class MessageReceivedEvent extends BaseEvent {
  constructor(agentId: AgentId, sessionId: SessionId, public readonly contentPreview: string) {
    super("message_received", agentId, sessionId);
  }
}

// 4. ToolCallRequestedEvent
export class ToolCallRequestedEvent extends BaseEvent {
  constructor(agentId: AgentId, sessionId: SessionId, public readonly toolCall: ToolCall) {
    super("tool_call_requested", agentId, sessionId);
  }
}

// 5. ToolCallCompletedEvent
export class ToolCallCompletedEvent extends BaseEvent {
  constructor(
    agentId: AgentId,
    sessionId: SessionId,
    public readonly toolName: string,
    public readonly success: boolean,
    public readonly durationMs: number,
  ) {
    super("tool_call_completed", agentId, sessionId);
  }

  override toJSON(): BaseEventJSON & {
    tool_name: string;
    success: boolean;
    duration_ms: number;
  } {
    return {
      ...super.toJSON(),
      tool_name: this.toolName,
      success: this.success,
      duration_ms: this.durationMs,
    };
  }
}

// 6. TokenUsageUpdatedEvent
export class TokenUsageUpdatedEvent extends BaseEvent {
  constructor(agentId: AgentId, sessionId: SessionId, public readonly usage: TokenUsage) {
    super("token_usage_updated", agentId, sessionId);
  }
}

// 7. StreamChunkEvent
export class StreamChunkEvent extends BaseEvent {
  constructor(agentId: AgentId, sessionId: SessionId, public readonly content: string) {
    super("stream_chunk", agentId, sessionId);
  }
}

// 8. ErrorEvent
export class ErrorEvent extends BaseEvent {
  constructor(agentId: AgentId, sessionId: SessionId, public readonly errorMessage: string) {
    super("error", agentId, sessionId);
  }
}

// 9. GoalPhaseChangedEvent
export class GoalPhaseChangedEvent extends BaseEvent {
  constructor(agentId: AgentId, sessionId: SessionId, public readonly phase: string, public readonly cycle: number) {
    super("goal_phase_changed", agentId, sessionId);
  }
}

// 10. GoalTaskCompletedEvent
export class GoalTaskCompletedEvent extends BaseEvent {
  constructor(
    agentId: AgentId,
    sessionId: SessionId,
    public readonly taskId: string,
    public readonly taskDescription: string,
    public readonly success: boolean,
  ) {
    super("goal_task_completed", agentId, sessionId);
  }
}

// 11. GoalCycleCompleteEvent
export class GoalCycleCompleteEvent extends BaseEvent {
  constructor(
    agentId: AgentId,
    sessionId: SessionId,
    public readonly cycle: number,
    public readonly tasksCompleted: number,
    public readonly tasksFailed: number,
    public readonly verificationPassed: boolean,
  ) {
    super("goal_cycle_complete", agentId, sessionId);
  }
}

// 12. GuardrailDecisionEvent
export class GuardrailDecisionEvent extends BaseEvent {
  constructor(
    agentId: AgentId,
    sessionId: SessionId,
    public readonly phase: string,
    public readonly decision: string,
    public readonly reason: string,
    public readonly guardrailName: string,
  ) {
    super("guardrail_decision", agentId, sessionId);
  }

  override toJSON(): BaseEventJSON & {
    phase: string;
    decision: string;
    reason: string;
    guardrail_name: string;
  } {
    return {
      ...super.toJSON(),
      phase: this.phase,
      decision: this.decision,
      reason: this.reason,
      guardrail_name: this.guardrailName,
    };
  }
}

// ---------------------------------------------------------------------------
// EventHandler
// ---------------------------------------------------------------------------

/**
 * Subscriber contract for the EventBus. Implementations may return a Promise
 * (async handlers run concurrently per publish) or void (synchronous).
 * Exceptions are isolated per handler so one failure does not break others.
 */

export interface EventHandler {
  handle(ev: AgentEvent): Promise<void> | void;
  readonly name: string;
}

// ---------------------------------------------------------------------------
// EventBus
// ---------------------------------------------------------------------------

/**
 * Minimal in-process pub/sub used to fan agent lifecycle events out to
 * subscribers. Subscriptions are keyed by {@link EventHandler.name} so a
 * later registration with the same name overwrites the earlier one.
 * publish() awaits all handlers concurrently; per-handler failures are
 * routed to the installed ErrorHandler (default: console.error) and never
 * reject the publish() promise.
 */

export type ErrorHandler = (handler: string, event: AgentEvent, error: Error) => void;

export class EventBus {
  private _subs = new Map<string, EventHandler>();
  private _errorHandler: ErrorHandler | null = null;

  subscribe(handler: EventHandler): void {
    this._subs.set(handler.name, handler);
  }

  unsubscribe(id: string): void {
    this._subs.delete(id);
  }

  /** Set a custom error handler for handler failures. */
  setErrorHandler(handler: ErrorHandler): void {
    this._errorHandler = handler;
  }

  /**
   * Fan `ev` out to every subscriber. Handlers run concurrently via
   * Promise.all; a throwing handler is caught and reported to the error
   * handler rather than rejecting this promise, so a single misbehaving
   * subscriber cannot break the event stream.
   */
  async publish(ev: AgentEvent): Promise<void> {
    const handlers = [...this._subs.values()];
    const promises = handlers.map(async (h) => {
      try {
        await h.handle(ev);
      } catch (err) {
        const error = err instanceof Error ? err : new Error(String(err));
        if (this._errorHandler) {
          this._errorHandler(h.name, ev, error);
        } else {
          // Default: log to console for debugging
          console.error(`[EventBus] handler "${h.name}" failed for event "${ev.eventType}":`, error.message);
        }
      }
    });
    await Promise.all(promises);
  }

  get handlerCount(): number {
    return this._subs.size;
  }
}
